data_dev/driver/src/data_dev_top.c

```diff
@@ -67,234 +66,347 @@ MODULE_DEVICE_TABLE(pci, DataDev_Ids);
+   dev->cfgIrqHold    = cfgIrqHold;    // IRQ hold configuration
+   dev->cfgIrqDis     = cfgIrqDis;     // IRQ disable flag
+   dev->cfgBgThold[0] = cfgBgThold0;   // Background threshold 0
+   dev->cfgBgThold[1] = cfgBgThold1;   // Background threshold 1
+   dev->cfgBgThold[2] = cfgBgThold2;   // Background threshold 2
+   dev->cfgBgThold[3] = cfgBgThold3;   // Background threshold 3
+   dev->cfgBgThold[4] = cfgBgThold4;   // Background threshold 4
+   dev->cfgBgThold[5] = cfgBgThold5;   // Background threshold 5
+   dev->cfgBgThold[6] = cfgBgThold6;   // Background threshold 6
+   dev->cfgBgThold[7] = cfgBgThold7;   // Background threshold 7
+
+   // Assign the IRQ number from the pci_dev structure
    dev->irq = pcidev->irq;
 
-   // Set device fields
-   dev->pcidev = pcidev;
-   dev->device = &(pcidev->dev);
-   dev->hwFunc = hfunc;
+   // Set basic device context
+   dev->pcidev = pcidev;               // PCI device structure
+   dev->device = &(pcidev->dev);       // Device structure
+   dev->hwFunc = hfunc;                // Hardware function pointer
 
-   dev->reg    = dev->base + AGEN2_OFF;
-   dev->rwBase = dev->base + PHY_OFF;
-   dev->rwSize = (2*USER_SIZE) - PHY_OFF;
+   // Initialize device memory regions
+   dev->reg    = dev->base + AGEN2_OFF;   // Register base address
+   dev->rwBase = dev->base + PHY_OFF;     // Read/Write base address
+   dev->rwSize = (2*USER_SIZE) - PHY_OFF; // Read/Write region size
 
-   // Set and clear reset
+   // Manage device reset cycle
    dev_info(dev->device,"Init: Setting user reset\n");
-   AxiVersion_SetUserReset(dev->base + AVER_OFF,true);
+   AxiVersion_SetUserReset(dev->base + AVER_OFF,true); // Set user reset
    dev_info(dev->device,"Init: Clearing user reset\n");
-   AxiVersion_SetUserReset(dev->base + AVER_OFF,false);
-
-   // 128bit desc, = 64-bit address map
-   if ( (ioread32(dev->reg) & 0x10000) != 0) {
-      // Get the AXI Address width (in units of bits)
-      axiWidth = (ioread32(dev->reg+0x34) >> 8) & 0xFF;
-      if ( !dma_set_mask_and_coherent(dev->device, DMA_BIT_MASK(axiWidth)) ) {
-         dev_info(dev->device,"Init: Using %d-bit DMA mask.\n",axiWidth);
+   AxiVersion_SetUserReset(dev->base + AVER_OFF,false); // Clear user reset
+
+   // Configure DMA based on AXI address width: 128bit desc, = 64-bit address map
+   if ((readl(dev->reg) & 0x10000) != 0) {
+      axiWidth = (readl(dev->reg + 0x34) >> 8) & 0xFF; // Extract AXI address width
+
+      // Attempt to set DMA and coherent DMA masks based on AXI width
+      if (!dma_set_mask(dev->device, DMA_BIT_MASK(axiWidth))) {
+         dev_info(dev->device, "Init: Using %d-bit DMA mask.\n", axiWidth);
+
+         if (!dma_set_coherent_mask(dev->device, DMA_BIT_MASK(axiWidth))) {
+            dev_info(dev->device, "Init: Using %d-bit coherent DMA mask.\n", axiWidth);
+         } else {
+            dev_warn(dev->device, "Init: Failed to set coherent DMA mask.\n");
+         }
```

Comment:
Are failures here not considered fatal? Kernel docs say we should not attempt to use DMA until `dma_set_mask` returns success.

Reply:
I'll defer that question to @slacrherbst, who wrote the driver